### ISSUE TITLE
Regarding the Dead Code Elimination 

### ISSUE THREAD

**Issue by pdamoc** [dotnetCarpenter/elmtut#1]
Elm's compiler does DCE at the level of the function in Elm code. In other words, when using libraries and modules, it only pulls in the minimum amount of code. If you have dead code in Elm, it will never get into the output. 

The dead code you are seeing in JS is actually bits of the runtime that are implemented in a private JS API. The private JS code used to implement core functionality is not optimized for DCE. The amount of dead code there is usually very small and constant (it doesn't grow with your app). 

There are also explorations of further optimization of outputted JS code [see more here](https://dev.to/skinney/improving-elm-s-compiler-output-5e1h) but Elm is fast enough and we are entering into the land of diminishing returns.

**Comment by mfeineis** [comment]
Pretty much everybody uses some kind of minifier for their JS, Elm code is really nicely structured for this. For example the functions you've listed in your readme @dotnetCarpenter 

```js
var _Utils_notEqual = F2(function(a, b) { return !_Utils_eq(a,b); });
var _Utils_ge = F2(function(a, b) { return _Utils_cmp(a, b) >= 0; });
var _Utils_compare = F2(function(x, y)
var _Utils_Tuple0_UNUSED = { $: '#0' };
function _Utils_Tuple2_UNUSED(a, b) { return { $: '#2', a: a, b: b }; }
function _Utils_Tuple3(a, b, c) { return { a: a, b: b, c: c }; }
function _Utils_Tuple3_UNUSED(a, b, c) { return { $: '#3', a: a, b: b, c: c }; }
function _Utils_chr_UNUSED(c) { return new String(c); }
function _Utils_update(oldRecord, updatedFields)
var _List_Nil_UNUSED = { $: '[]' };
function _List_Cons_UNUSED(hd, tl) { return { $: '::', a: hd, b: tl }; }
function _Debug_todo(moduleName, region)
function _Debug_todoCase(moduleName, region, value)
function _Debug_toString_UNUSED(value)
function _Debug_crash_UNUSED(identifier, fact1, fact2, fact3, fact4)
```
will be removed when unused by whatever you use, see also [the minification section of the guide](https://guide.elm-lang.org/optimization/asset_size.html) although I'd recommend using [terser](https://github.com/terser/terser) since [uglify will not support ES6+ syntax for the forseeable future](https://github.com/mishoo/UglifyJS2#note)

**Comment by pdamoc** [comment]
> For example the functions you've listed in your readme  ...  will be removed 

That's not actually true. Yes, the minifiers reduce the size a lot but do not remove these functions. Please try `terser` without mangle and search for the above names of functions.

**Comment by dotnetCarpenter** [comment]
@mfeineis  I have a [Makefile](../blob/master/forms/Makefile) that does the minification described in https://github.com/elm/compiler/blob/master/hints/optimize.md#instructions. I think `terser` and `uglify` will yield similar results, since Elm only outputs ES5.

As @pdamoc said, it does not remove dead code. Only the instances where the compiler output stuff like:
```js
/**_UNUSED/
if (x.$ === 'Set_elm_builtin')
{
	x = $elm$core$Set$toList(x);
	y = $elm$core$Set$toList(y);
}
if (x.$ === 'RBNode_elm_builtin' || x.$ === 'RBEmpty_elm_builtin')
{
	x = $elm$core$Dict$toList(x);
	y = $elm$core$Dict$toList(y);
}
//*/
```

You can check what the result of running `elm make --optmize` in [main.js](../blob/master/forms/main.js) and the result of the minification in [main.min.js](../blob/master/forms/main.min.js).

If you use Chromium based devtools (I use Opera) you have a **Coverage** tab, hidden in Menu (3 vertical dots) -> More tools -> Coverage. 

If I check index.html after running `elm make src/Main.elm`, I see that my unused code is gone from the output.
If I check index.html and reference `main.js`, which was generated with `elm make --optimized --output main.js src/Main.elm`, I find the exact same Elm code as when I did not use `--optimized`.

I have use the same input, **12345678aB** in 3 different settings. The first being the normal Elm app. The second with the optimized flag and the last with minification. Overall, the percentage of unused code almost identical in all 3 instances. In fact it gets a little worse with the optimized flag.
The different results are:
```
http://localhost:8000/index.html
114 090 bytes
unused: 60 365 bytes 52.9% 

http://localhost:8000/index2.html
Total: 112 342 bytes
unused: 59 539 bytes 53.0% 

http://localhost:8000/index2.html (minified)
Total: 28 975 bytes
unused: 14 697 bytes 50.7% 
```

These findings match with what @pdamoc explained.
> The dead code you are seeing in JS is actually bits of the runtime that are implemented in a private JS API. The private JS code used to implement core functionality is not optimized for DCE. The amount of dead code there is usually very small and constant (it doesn't grow with your app).

To play with it yourself, you could clone this repo and `cd` into the forms folder and type `make` - it should create all of the different assets that I have talked about (unless I made a mistake in the Makefile).

Be aware though that `elm reactor` or `snap` caches the `.js` files. You might need to use a different server, if you change them. Personally I use ` python3 -m http.server`.

**Comment by dotnetCarpenter** [comment]
> **we are entering into the land of diminishing returns.**

@pdamoc I agree but the result when just starting with Elm, after reading all those articles about DCE and how small Elm is compared to other frameworks, is shocking. I would love to sprinkle existing websites with Elm code, but having a big constant chunk of dead code (around 50%) at each location is discouraging.

It is probably not 50% in real usage.. I just have not gotten far enough with Elm and have not produced any real production code to include in a website. So I am not hugely afraid. Just a little shocked. It went against all of my expectations but after taking some time to consider, it is not a deal breaker at all.

That said, I might try to get involved with the optimization effort for micro Elm projects. Thank you for the article! I will read it later.

**Comment by mfeineis** [comment]
> > For example the functions you've listed in your readme  ...  will be removed
> 
> That's not actually true. Yes, the minifiers reduce the size a lot but do not remove these functions. Please try `terser` without mangle and search for the above names of functions.

I made [a simple Hello World](https://github.com/mfeineis/elm-0.19.1-hello-minifier) with the setup I usually use, none of the functions listed are in the [unmangled output](https://raw.githubusercontent.com/mfeineis/elm-0.19.1-hello-minifier/master/dist/app.optimized.terser1.unmangled.js). I think the `NODE_ENV=production` does some extra magic.

The complete app is about 7kb gzipped, not a big hurdle to sprinkling in some Elm in your apps if you ask me.

**Comment by dotnetCarpenter** [comment]
@mfeineis I appreciate you take the time to put up a repo with your compiler options and `main` function. For me it's very interesting to see. But I also feel that I need to learn a lot more about Elm before I venture into other peoples projects.

I did a clone and the initial numbers show the same as my numbers. Elm does not do Dead Code Elimination (DCE) on its run time. It is only the Elm code that you write that gets DCE.
`terser` does not seem to do any DCE, regardless of environment variable flag.

If DCE was enabled on all code that Elm generates, the coverage would be close to or a 100%, once we have used every code path in our program.

Your HelloWorld example with `dist/app.optimized.terser2.mangled.js` has a coverage of 54.1%. The unoptimized version `dist/app.default.terser2.mangled.js` has a coverage of 40.3%.

If we use the web server that comes with `elm reactor`, Snap/1.1.1.1, then we get gzip on all `.js` files.
Measuring index.html with `dist/app.optimized.terser2.mangled.js` give us:
 app.optimized.terser2.mangled.js size is 6,5kb (gzipped) and 15.8kb after being uncompressed.

To fully understand what the cost of JS is, we have to appreciate what the browser does with the JS. In this case, it downloads 6.5kb over the wire, then it uncompress the file, then it begins to parse the JS into AST and then it begins to run the code in a VM, while JIT compiling hot functions in the background or at intervals.

Imagine we had the perfect compiler. Your HelloWorld example would compile to:
```js
'use strict';var Elm={};Elm.Main={};Elm.Main.init=function(o){o.node.textContent='Hello, World!'}
```
This has 100% coverage and is 370b (gzipped) and 97b after being uncompressed. Gzipping has an overhead. Which with extremely small files gets bigger than the actual file.

My point is, there is quite a few things happening after you have downloaded JS and I don't want to add **duplicated** code in the ballpark of 6-7000 bytes, that needs to be uncompressed and parsed, jitted etc, if I do not have to. For small stuff, I'll write a few 100 lines of JS and be happy with that.

But @pdamoc is right, the more Elm code you have, the less does the runtime size matter and the runtime is much smaller than vue.js or react or even just Immutable.js.

Let's just be honest here. Elm does not apply DCE to its runtime and that is very far from obvious when you read the blog posts and the documentation. It seems that most of us got fooled but it is really not a big deal.

**Comment by pdamoc** [comment]
@mfeineis, my bad, I did some more testing and it looks like I need to investigate further my production pipeline. The functions get removed on raw compiler outputs. I was looking at some parcel output and it looks like that breaks the tree shaking.

**Comment by pdamoc** [comment]
> I don't want to add duplicated code in the ballpark of 6-7000 bytes

If you have multiple widgets on a page, you can bundle them all into a single JS file that shares the runtime. This way it will not get duplicated.

**Comment by pdamoc** [comment]
@mfeineis, I found out my error, I was looking at terser minification without pure_funcs. Once pure_funcs is added, the unused functions are removed. 

@dotnetCarpenter, one more thing about the code that still remains after minification. most of the code you see there is the code from the virtual dom and the JSON decoders needed for the ports. In theory, this can be further optimized by the compiler if the type of the main is `Html msg` or if the flags' type is `()` .

**Comment by mfeineis** [comment]
@dotnetCarpenter yes, let's chat again after you've finished your guide walkthrough. I'm pretty interested in what beginner problems using Elm are.

The DCE certainly isn't at a level it could be, that's for sure. Like pdamoc said, the `--optimize` mostly takes care of shortening record member names which is not easy for external tools to do without breaking stuff. The numbers are also compared to other JS libraries: if you can do something in 100 lines of vanilla JS, that's probably not where should bring out the React/Angular/Elm hammer. But if you have a fashionable SPA the number features per byte Elm gives you compared to others is extremely reasonable. I'm not saying going that route is always the best approach, though.

**Comment by dotnetCarpenter** [comment]
@mfeineis I was wrong about `terser` and DCE. It does DCE with the `--compress` flag. I'll switch to `terser` and shamelessly use your build configuration 🙂
